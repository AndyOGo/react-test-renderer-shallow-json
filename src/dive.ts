import type { ReactTestRendererTreeNodeFixed } from './types';

export function dive(
  tree: ReactTestRendererTreeNodeFixed,
  depth: number
): number {
  return typeof tree === 'string' || tree.nodeType === 'host'
    ? depth
    : depth - 1;
}
