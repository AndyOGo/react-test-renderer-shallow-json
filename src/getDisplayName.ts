import {
  ReactTestRendererTreeFixed,
  isContext,
  isNamedExoticComponentType,
} from './types';

const elementSymbolMap: Readonly<Record<symbol, string>> = {
  [Symbol.for('react.memo')]: 'Memo',
  [Symbol.for('react.context')]: 'Consumer',
  [Symbol.for('react.provider')]: 'Provider',
};

function appendSymbol(displayName: string, $$typeof?: symbol): string {
  if (!$$typeof || !elementSymbolMap[$$typeof]) {
    return displayName;
  }

  return `${displayName}.${elementSymbolMap[$$typeof]}`;
}

export function getDisplayName(
  type: ReactTestRendererTreeFixed['type'],
  fallback = '???'
): string {
  if (typeof type === 'string') {
    return type;
  }

  if (isContext(type)) {
    return appendSymbol(type._context.displayName || fallback, type.$$typeof);
  }

  if (isNamedExoticComponentType(type)) {
    return (
      type.displayName ||
      appendSymbol(
        type.type ? getDisplayName(type.type) : fallback,
        type.$$typeof
      )
    );
  }

  return type.displayName || type.name || fallback;
}
