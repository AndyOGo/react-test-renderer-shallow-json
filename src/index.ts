console.log('Start coding here...');
