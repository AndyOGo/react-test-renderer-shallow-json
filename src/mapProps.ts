import type { ReactTestRendererTreeFixed } from './types';

export function mapProps(props: ReactTestRendererTreeFixed['props']) {
  const { children, ...propsWithoutChildren } = props;

  return propsWithoutChildren;
}
