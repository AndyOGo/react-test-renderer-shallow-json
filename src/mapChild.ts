import type {
  ReactTestRendererNodeFixed,
  ReactTestRendererTreeChild,
} from './types';
import { getDisplayName } from './getDisplayName';
import { mapProps } from './mapProps';
import { mapChildren } from './mapChildren';

export function mapChild(
  child: ReactTestRendererTreeChild
): ReactTestRendererNodeFixed {
  if (typeof child === 'string') {
    return child;
  }

  const { key, ref } = child;

  const json = {
    type: getDisplayName(child.type),
    props: mapProps({
      ...child.props,
      ...(key && { key }),
      ...(ref && { ref }),
    }),
    children: mapChildren(child.props),
  };

  Object.defineProperty(json, '$$typeof', {
    value: Symbol.for('react.test.json'),
  });

  return json;
}
