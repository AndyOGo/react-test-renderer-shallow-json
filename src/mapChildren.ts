import type {
  ReactTestRendererJSONFixed,
  ReactTestRendererTreeFixed,
} from './types';
import { mapChild } from './mapChild';

export function mapChildren(
  props: ReactTestRendererTreeFixed['props']
): ReactTestRendererJSONFixed['children'] {
  const { children } = props;

  if (!children) {
    return null;
  }

  if (!Array.isArray(children)) {
    return [mapChild(children)];
  }

  return children.map(mapChild);
}
