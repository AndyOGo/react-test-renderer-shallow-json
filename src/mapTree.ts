import type { ReactTestRendererNode } from 'react-test-renderer';
import type { ReactTestRendererTreeNodeFixed } from './types';
import { getDisplayName } from './getDisplayName';
import { mapProps } from './mapProps';
import { mapChildren } from './mapChildren';
import { mapTrees } from './mapTrees';
import { withReactTestSymbol } from './withReactTestSymbol';

export function mapTree(
  tree: ReactTestRendererTreeNodeFixed,
  depth = 1
): ReactTestRendererNode {
  if (typeof tree === 'string') {
    return tree;
  }

  const json = {
    type: getDisplayName(tree.type),
    props: mapProps(tree.props),
    children:
      depth > 0 ? mapTrees(tree.rendered, depth) : mapChildren(tree.props),
  };

  return withReactTestSymbol(json) as ReactTestRendererNode;
}
