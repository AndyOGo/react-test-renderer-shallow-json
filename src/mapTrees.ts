import type { ReactTestRendererJSON } from 'react-test-renderer';
import type { ReactTestRendererTreeFixed } from './types';
import { mapTree } from './mapTree';
import { dive } from './dive';

export function mapTrees(
  trees: ReactTestRendererTreeFixed['rendered'],
  depth = 1
): ReactTestRendererJSON['children'] {
  if (!trees) {
    return null;
  }

  if (Array.isArray(trees)) {
    return trees.map((tree) => mapTree(tree, dive(tree, depth)));
  }

  return [mapTree(trees, dive(trees, depth))];
}
