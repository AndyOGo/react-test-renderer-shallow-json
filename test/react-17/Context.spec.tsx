import React from 'react';
import { create } from 'react-test-renderer';
import { shallowJSON } from '../../src';
import {
  Context,
  ContextConsumer,
  ContextProvider,
  NamedContext,
  NamedContextConsumer,
  NamedContextProvider,
} from './Context';
import { TestWrapper } from './TestWrapper';

describe('Context', () => {
  it('matches snapshot', () => {
    expect(
      shallowJSON(
        create(
          <Context.Provider value="provided">
            <ContextConsumer />
          </Context.Provider>
        ).toTree()
      )
    ).toMatchSnapshot();
  });

  it('matches snapshot as children', () => {
    expect(
      shallowJSON(
        create(
          <TestWrapper>
            <Context.Provider value="provided">
              <ContextConsumer />
            </Context.Provider>
          </TestWrapper>
        ).toTree()
      )
    ).toMatchSnapshot();
  });
});

describe('ContextProvider', () => {
  it('matches snapshot', () => {
    expect(
      shallowJSON(
        create(
          <ContextProvider>
            <ContextConsumer />
          </ContextProvider>
        ).toTree(),
        { depth: 2 }
      )
    ).toMatchSnapshot();
  });

  it('matches snapshot as children', () => {
    expect(
      shallowJSON(
        create(
          <TestWrapper>
            <ContextProvider>
              <ContextConsumer />
            </ContextProvider>
          </TestWrapper>
        ).toTree()
      )
    ).toMatchSnapshot();
  });
});

describe('NamedContext', () => {
  it('matches snapshot', () => {
    expect(
      shallowJSON(
        create(
          <NamedContext.Provider value="provided">
            <NamedContextConsumer />
          </NamedContext.Provider>
        ).toTree()
      )
    ).toMatchSnapshot();
  });

  it('matches snapshot as children', () => {
    expect(
      shallowJSON(
        create(
          <TestWrapper>
            <NamedContext.Provider value="provided">
              <NamedContextConsumer />
            </NamedContext.Provider>
          </TestWrapper>
        ).toTree()
      )
    ).toMatchSnapshot();
  });
});

describe('NamedContextProvider', () => {
  it('matches snapshot', () => {
    expect(
      shallowJSON(
        create(
          <NamedContextProvider>
            <NamedContextConsumer />
          </NamedContextProvider>
        ).toTree(),
        { depth: 2 }
      )
    ).toMatchSnapshot();
  });

  it('matches snapshot as children', () => {
    expect(
      shallowJSON(
        create(
          <TestWrapper>
            <NamedContextProvider>
              <NamedContextConsumer />
            </NamedContextProvider>
          </TestWrapper>
        ).toTree()
      )
    ).toMatchSnapshot();
  });
});
