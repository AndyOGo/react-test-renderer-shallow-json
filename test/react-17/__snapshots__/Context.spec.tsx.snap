// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context matches snapshot 1`] = `
<ContextConsumer>
  provided
</ContextConsumer>
`;

exports[`Context matches snapshot as children 1`] = `
<TestWrapper>
  <div>
    <Context.Provider
      value="provided"
    >
      <ContextConsumer />
    </Context.Provider>
  </div>
</TestWrapper>
`;

exports[`ContextProvider matches snapshot 1`] = `
<ContextProvider>
  <ContextConsumer />
</ContextProvider>
`;

exports[`ContextProvider matches snapshot as children 1`] = `
<TestWrapper>
  <div>
    <ContextProvider>
      <ContextConsumer />
    </ContextProvider>
  </div>
</TestWrapper>
`;

exports[`NamedContext matches snapshot 1`] = `
<NamedContextConsumer>
  provided
</NamedContextConsumer>
`;

exports[`NamedContext matches snapshot as children 1`] = `
<TestWrapper>
  <div>
    <NamedContextDisplayName.Provider
      value="provided"
    >
      <NamedContextConsumer />
    </NamedContextDisplayName.Provider>
  </div>
</TestWrapper>
`;

exports[`NamedContextProvider matches snapshot 1`] = `
<NamedContextProvider>
  <NamedContextConsumer />
</NamedContextProvider>
`;

exports[`NamedContextProvider matches snapshot as children 1`] = `
<TestWrapper>
  <div>
    <NamedContextProvider>
      <NamedContextConsumer />
    </NamedContextProvider>
  </div>
</TestWrapper>
`;
